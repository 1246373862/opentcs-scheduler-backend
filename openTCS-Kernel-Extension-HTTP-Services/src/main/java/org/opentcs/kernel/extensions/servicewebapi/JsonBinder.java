/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import java.io.IOException;

/**
 * Binds JSON strings to objects and vice versa.
 */
public class JsonBinder {

  /**
   * Maps between objects and their JSON representations.
   */
  private final ObjectMapper objectMapper
      = new ObjectMapper()
          .registerModule(new JavaTimeModule())
          .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

  /**
   * Creates a new instance.
   */
  public JsonBinder() {
  }

  /**
   * Maps the given JSON string to an object.
   *
   * @param <T> The type of object to map to.
   * @param jsonString The JSON string.
   * @param clazz The type of object to map to.
   * @return The object created from the JSON string.
   * @throws IllegalArgumentException In case there was a problem mapping the given object from
   * JSON.
   * (An IllegalArgumentException is mapped to HTTP status code 400, indicating a client error.)
   */
  public <T> T fromJson(String jsonString, Class<T> clazz)
      throws IllegalArgumentException {
    try {
      return objectMapper.readValue(jsonString, clazz);
    }
    catch (IOException exc) {
      throw new IllegalArgumentException("Could not parse JSON input", exc);
    }
  }

  /**
   * Maps the given object to a JSON string.
   *
   * @param object The object to be mapped.
   * @return The JSON string representation of the object.
   * @throws IllegalStateException In case there was a problem mapping the given object to JSON.
   * (An IllegalStateException is mapped to HTTP status code 500, indicating an internal error.)
   */
  public String toJson(Object object)
      throws IllegalStateException {
    try {
      return objectMapper
          .writerWithDefaultPrettyPrinter()
          .writeValueAsString(object);
    }
    catch (JsonProcessingException exc) {
      throw new IllegalStateException("Could not produce JSON output", exc);
    }
  }

  /**
   * Maps the given throwable to a JSON string.
   *
   * @param t The throwable to be mapped.
   * @return A JSON string for the given throwable, consisting of a single-element array containing
   * the throwable's message.
   * @throws IllegalStateException In case there was a problem mapping the given object to JSON.
   * (An IllegalStateException is mapped to HTTP status code 500, indicating an internal error.)
   */
  public String toJson(Throwable t)
      throws IllegalStateException {
    try {
      return objectMapper
          .writerWithDefaultPrettyPrinter()
          .writeValueAsString(objectMapper.createArrayNode().add(t.getMessage()));
    }
    catch (JsonProcessingException exc) {
      throw new IllegalStateException("Could not produce JSON output", exc);
    }
  }

}
