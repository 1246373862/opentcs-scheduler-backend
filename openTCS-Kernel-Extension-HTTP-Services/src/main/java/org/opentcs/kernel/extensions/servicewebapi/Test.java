package org.opentcs.kernel.extensions.servicewebapi;

import org.opentcs.components.kernel.KernelExtension;

public class Test implements KernelExtension {

  private volatile boolean enabled;

  @Override
  public void initialize() {
    if (enabled){
      return;
    }
    System.out.println("初始化~~~~~初始化~~~~~初始化~~~~~初始化~~~~~初始化~~~~~初始化~~~~~");
  }

  @Override
  public boolean isInitialized() {
    return enabled;
  }

  @Override
  public void terminate() {
    if (!enabled) {
      return;
    }
    enabled = false;
    System.out.println("结束！！");
  }
}
