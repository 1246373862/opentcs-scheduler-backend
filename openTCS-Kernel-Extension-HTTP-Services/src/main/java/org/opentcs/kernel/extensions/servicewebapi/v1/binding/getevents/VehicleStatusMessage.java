/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import static java.util.Objects.requireNonNull;
import java.util.Set;
import java.util.stream.Collectors;
import org.opentcs.data.model.TCSResourceReference;
import org.opentcs.data.model.Vehicle;

/**
 * A status message containing information about a vehicle.
 */
public class VehicleStatusMessage
    extends StatusMessage {

  private String vehicleName = "";

  private String transportOrderName = "";

  private String position;

  private PrecisePosition precisePosition;

  private boolean paused;

  private Vehicle.State state;

  private Vehicle.ProcState procState;

  private List<List<String>> allocatedResources = new ArrayList<>();

  private List<List<String>> claimedResources = new ArrayList<>();

  /**
   * Creates a new instance.
   */
  public VehicleStatusMessage() {
  }

  @Override
  public VehicleStatusMessage setSequenceNumber(long sequenceNumber) {
    return (VehicleStatusMessage) super.setSequenceNumber(sequenceNumber);
  }

  @Override
  public VehicleStatusMessage setCreationTimeStamp(Instant creationTimeStamp) {
    return (VehicleStatusMessage) super.setCreationTimeStamp(creationTimeStamp);
  }

  public String getVehicleName() {
    return vehicleName;
  }

  public VehicleStatusMessage setVehicleName(String vehicleName) {
    this.vehicleName = vehicleName;
    return this;
  }

  public String getTransportOrderName() {
    return transportOrderName;
  }

  public VehicleStatusMessage setTransportOrderName(String transportOrderName) {
    this.transportOrderName = transportOrderName;
    return this;
  }

  public String getPosition() {
    return position;
  }

  public VehicleStatusMessage setPosition(String position) {
    this.position = position;
    return this;
  }

  public PrecisePosition getPrecisePosition() {
    return precisePosition;
  }

  public VehicleStatusMessage setPrecisePosition(PrecisePosition precisePosition) {
    this.precisePosition = precisePosition;
    return this;
  }

  public boolean isPaused() {
    return paused;
  }

  public VehicleStatusMessage setPaused(boolean paused) {
    this.paused = paused;
    return this;
  }

  public Vehicle.State getState() {
    return state;
  }

  public VehicleStatusMessage setState(Vehicle.State state) {
    this.state = state;
    return this;
  }

  public Vehicle.ProcState getProcState() {
    return procState;
  }

  public VehicleStatusMessage setProcState(Vehicle.ProcState procState) {
    this.procState = procState;
    return this;
  }

  public List<List<String>> getAllocatedResources() {
    return allocatedResources;
  }

  public VehicleStatusMessage setAllocatedResources(List<List<String>> allocatedResources) {
    this.allocatedResources = requireNonNull(allocatedResources, "allocatedResources");
    return this;
  }

  public List<List<String>> getClaimedResources() {
    return claimedResources;
  }

  public VehicleStatusMessage setClaimedResources(List<List<String>> claimedResources) {
    this.claimedResources = requireNonNull(claimedResources, "claimedResources");
    return this;
  }

  public static VehicleStatusMessage fromVehicle(Vehicle vehicle,
                                                 long sequenceNumber) {
    return fromVehicle(vehicle, sequenceNumber, Instant.now());
  }

  public static VehicleStatusMessage fromVehicle(Vehicle vehicle,
                                                 long sequenceNumber,
                                                 Instant creationTimeStamp) {
    VehicleStatusMessage vehicleMessage = new VehicleStatusMessage();
    vehicleMessage.setSequenceNumber(sequenceNumber);
    vehicleMessage.setCreationTimeStamp(creationTimeStamp);
    vehicleMessage.setVehicleName(vehicle.getName());
    vehicleMessage.setTransportOrderName(
        vehicle.getTransportOrder() == null ? null : vehicle.getTransportOrder().getName());
    vehicleMessage.setPosition(
        vehicle.getCurrentPosition() == null ? null : vehicle.getCurrentPosition().getName());
    vehicleMessage.setPaused(vehicle.isPaused());
    vehicleMessage.setState(vehicle.getState());
    vehicleMessage.setProcState(vehicle.getProcState());
    if (vehicle.getPrecisePosition() != null) {
      vehicleMessage.setPrecisePosition(new PrecisePosition(vehicle.getPrecisePosition().getX(),
                                                            vehicle.getPrecisePosition().getY(),
                                                            vehicle.getPrecisePosition().getZ()));
    }
    vehicleMessage.setAllocatedResources(toListOfListOfNames(vehicle.getAllocatedResources()));
    vehicleMessage.setClaimedResources(toListOfListOfNames(vehicle.getClaimedResources()));
    return vehicleMessage;
  }

  private static List<List<String>> toListOfListOfNames(
      List<Set<TCSResourceReference<?>>> resources) {
    List<List<String>> result = new ArrayList<>(resources.size());

    for (Set<TCSResourceReference<?>> resSet : resources) {
      result.add(
          resSet.stream()
              .map(resRef -> resRef.getName())
              .collect(Collectors.toList())
      );
    }

    return result;
  }

  /**
   * A precise position of a vehicle.
   */
  public static class PrecisePosition {

    private long x;

    private long y;

    private long z;

    /**
     * Creates a new instance.
     */
    public PrecisePosition() {
    }

    /**
     * Creates a new instance.
     *
     * @param x x value
     * @param y y value
     * @param z z value
     */
    public PrecisePosition(long x, long y, long z) {
      this.x = x;
      this.y = y;
      this.z = z;
    }

    public long getX() {
      return x;
    }

    public void setX(long x) {
      this.x = x;
    }

    public long getY() {
      return y;
    }

    public void setY(long y) {
      this.y = y;
    }

    public long getZ() {
      return z;
    }

    public void setZ(long z) {
      this.z = z;
    }
  }
}
