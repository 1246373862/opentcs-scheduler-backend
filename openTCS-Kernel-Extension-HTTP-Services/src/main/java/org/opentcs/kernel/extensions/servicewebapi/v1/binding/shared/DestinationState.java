/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import static java.util.Objects.requireNonNull;
import javax.annotation.Nonnull;
import org.opentcs.data.order.DriveOrder;

/**
 * A {@link org.opentcs.data.order.DriveOrder DriveOrder}'s destination.
 */
public class DestinationState {

  private String locationName = "";

  private String operation = "";

  private State state = State.PRISTINE;

  private List<Property> properties = new ArrayList<>();

  /**
   * Creates a new instance.
   */
  public DestinationState() {
  }

  @Nonnull
  public String getLocationName() {
    return locationName;
  }

  public DestinationState setLocationName(@Nonnull String name) {
    this.locationName = requireNonNull(name, "name");
    return this;
  }

  @Nonnull
  public String getOperation() {
    return operation;
  }

  public DestinationState setOperation(@Nonnull String operation) {
    this.operation = requireNonNull(operation, "operation");
    return this;
  }

  @Nonnull
  public State getState() {
    return state;
  }

  public DestinationState setState(@Nonnull State state) {
    this.state = requireNonNull(state, "state");
    return this;
  }

  @Nonnull
  public List<Property> getProperties() {
    return properties;
  }

  public DestinationState setProperties(@Nonnull List<Property> properties) {
    this.properties = requireNonNull(properties, "properties");
    return this;
  }

  public static DestinationState fromDriveOrder(DriveOrder driveOrder) {
    if (driveOrder == null) {
      return null;
    }
    DestinationState destination = new DestinationState();
    destination.setLocationName(driveOrder.getDestination().getDestination().getName());
    destination.setOperation(driveOrder.getDestination().getOperation());
    destination.setState(mapDriveOrderState(driveOrder.getState()));

    for (Map.Entry<String, String> mapEntry
             : driveOrder.getDestination().getProperties().entrySet()) {
      destination.getProperties().add(new Property(mapEntry.getKey(), mapEntry.getValue()));
    }
    return destination;
  }

  private static DestinationState.State mapDriveOrderState(DriveOrder.State driveOrderState) {
    switch (driveOrderState) {
      case PRISTINE:
        return DestinationState.State.PRISTINE;
      case TRAVELLING:
        return DestinationState.State.TRAVELLING;
      case OPERATING:
        return DestinationState.State.OPERATING;
      case FINISHED:
        return DestinationState.State.FINISHED;
      case FAILED:
        return DestinationState.State.FAILED;
      default:
        throw new IllegalArgumentException("Unhandled drive order state: " + driveOrderState);
    }
  }

  /**
   * This enumeration defines the various states a DriveOrder may be in.
   */
  public enum State {

    /**
     * A DriveOrder's initial state, indicating it being still untouched/not
     * being processed.
     */
    PRISTINE,
    /**
     * Indicates this drive order being processed at the moment.
     */
    TRAVELLING,
    /**
     * Indicates the vehicle processing an order is currently executing an
     * operation.
     */
    OPERATING,
    /**
     * Marks a DriveOrder as successfully completed.
     */
    FINISHED,
    /**
     * Marks a DriveOrder as failed.
     */
    FAILED;
  }
}
