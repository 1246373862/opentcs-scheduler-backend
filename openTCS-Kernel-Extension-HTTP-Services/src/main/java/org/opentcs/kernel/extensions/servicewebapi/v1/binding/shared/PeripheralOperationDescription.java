/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import org.opentcs.data.peripherals.PeripheralOperation;
import org.opentcs.data.peripherals.PeripheralOperation.ExecutionTrigger;

/**
 * Describes a peripheral operation.
 */
public class PeripheralOperationDescription {

  private String operation;

  private String locationName;

  private ExecutionTrigger executionTrigger;

  private boolean completionRequired;

  public PeripheralOperationDescription() {
  }

  public String getOperation() {
    return operation;
  }

  public PeripheralOperationDescription setOperation(String operation) {
    this.operation = operation;
    return this;
  }

  public String getLocationName() {
    return locationName;
  }

  public PeripheralOperationDescription setLocationName(String locationName) {
    this.locationName = locationName;
    return this;
  }

  public ExecutionTrigger getExecutionTrigger() {
    return executionTrigger;
  }

  public PeripheralOperationDescription setExecutionTrigger(ExecutionTrigger executionTrigger) {
    this.executionTrigger = executionTrigger;
    return this;
  }

  public boolean isCompletionRequired() {
    return completionRequired;
  }

  public PeripheralOperationDescription setCompletionRequired(boolean completionRequired) {
    this.completionRequired = completionRequired;
    return this;
  }

  public static PeripheralOperationDescription fromPeripheralOperation(
      PeripheralOperation operation) {
    PeripheralOperationDescription state = new PeripheralOperationDescription();
    state.operation = operation.getOperation();
    state.locationName = operation.getLocation().getName();
    state.executionTrigger = operation.getExecutionTrigger();
    state.completionRequired = operation.isCompletionRequired();
    return state;
  }

}
