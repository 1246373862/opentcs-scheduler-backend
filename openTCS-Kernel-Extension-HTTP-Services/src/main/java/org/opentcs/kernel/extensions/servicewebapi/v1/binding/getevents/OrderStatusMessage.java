/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.opentcs.data.order.DriveOrder;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.DestinationState;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.Property;

/**
 * A status message containing details about a transport order.
 */
public class OrderStatusMessage
    extends StatusMessage {

  private String orderName;

  private String processingVehicleName;

  private OrderState orderState;

  private List<DestinationState> destinations = new ArrayList<>();

  private List<Property> properties = new ArrayList<>();

  /**
   * Creates a new instance.
   */
  public OrderStatusMessage() {
  }

  @Override
  public OrderStatusMessage setSequenceNumber(long sequenceNumber) {
    return (OrderStatusMessage) super.setSequenceNumber(sequenceNumber);
  }

  @Override
  public OrderStatusMessage setCreationTimeStamp(Instant creationTimeStamp) {
    return (OrderStatusMessage) super.setCreationTimeStamp(creationTimeStamp);
  }

  public String getOrderName() {
    return orderName;
  }

  public OrderStatusMessage setOrderName(String orderName) {
    this.orderName = orderName;
    return this;
  }

  public String getProcessingVehicleName() {
    return processingVehicleName;
  }

  public OrderStatusMessage setProcessingVehicleName(String processingVehicleName) {
    this.processingVehicleName = processingVehicleName;
    return this;
  }

  public OrderState getOrderState() {
    return orderState;
  }

  public OrderStatusMessage setOrderState(OrderState orderState) {
    this.orderState = orderState;
    return this;
  }

  public List<DestinationState> getDestinations() {
    return destinations;
  }

  public OrderStatusMessage setDestinations(List<DestinationState> destinations) {
    this.destinations = destinations;
    return this;
  }

  public List<Property> getProperties() {
    return properties;
  }

  public OrderStatusMessage setProperties(List<Property> properties) {
    this.properties = properties;
    return this;
  }

  public static OrderStatusMessage fromTransportOrder(TransportOrder order,
                                                      long sequenceNumber) {
    return fromTransportOrder(order, sequenceNumber, Instant.now());
  }

  public static OrderStatusMessage fromTransportOrder(TransportOrder order,
                                                      long sequenceNumber,
                                                      Instant creationTimeStamp) {
    OrderStatusMessage orderMessage = new OrderStatusMessage();
    orderMessage.setSequenceNumber(sequenceNumber);
    orderMessage.setCreationTimeStamp(creationTimeStamp);
    orderMessage.setOrderName(order.getName());
    orderMessage.setProcessingVehicleName(
        order.getProcessingVehicle() == null ? null : order.getProcessingVehicle().getName());
    orderMessage.setOrderState(OrderState.fromTransportOrderState(order.getState()));
    for (DriveOrder curDriveOrder : order.getAllDriveOrders()) {
      orderMessage.getDestinations().add(DestinationState.fromDriveOrder(curDriveOrder));
    }
    for (Map.Entry<String, String> mapEntry : order.getProperties().entrySet()) {
      orderMessage.getProperties().add(new Property(mapEntry.getKey(), mapEntry.getValue()));
    }
    return orderMessage;
  }

  /**
   * The various states a transport order may be in.
   */
  public enum OrderState {
    /**
     * A transport order's initial state.
     */
    RAW,
    /**
     * Indicates a transport order's parameters have been set up completely and the kernel should
     * dispatch it when possible.
     */
    ACTIVE,
    /**
     * Marks a transport order as ready to be dispatched to a vehicle.
     */
    DISPATCHABLE,
    /**
     * Marks a transport order as being processed by a vehicle.
     */
    BEING_PROCESSED,
    /**
     * Indicates the transport order is withdrawn from a processing vehicle but not yet in its
     * final state, as the vehicle has not yet finished/cleaned up.
     */
    WITHDRAWN,
    /**
     * Marks a transport order as successfully completed.
     */
    FINISHED,
    /**
     * General failure state that marks a transport order as failed.
     */
    FAILED,
    /**
     * Failure state that marks a transport order as unroutable.
     */
    UNROUTABLE;

    /**
     * Maps a transpor order's {@link TransportOrder#state state} to the corresponding
     * {@link OrderState}.
     *
     * @param state The transport order's state.
     * @return The corresponding OrderState.
     */
    public static OrderState fromTransportOrderState(TransportOrder.State state) {
      switch (state) {
        case RAW:
          return RAW;
        case ACTIVE:
          return ACTIVE;
        case DISPATCHABLE:
          return DISPATCHABLE;
        case BEING_PROCESSED:
          return BEING_PROCESSED;
        case WITHDRAWN:
          return WITHDRAWN;
        case FINISHED:
          return FINISHED;
        case FAILED:
          return FAILED;
        case UNROUTABLE:
          return UNROUTABLE;
        default:
          throw new IllegalArgumentException("Unknown transport order state.");
      }
    }
  }
}
