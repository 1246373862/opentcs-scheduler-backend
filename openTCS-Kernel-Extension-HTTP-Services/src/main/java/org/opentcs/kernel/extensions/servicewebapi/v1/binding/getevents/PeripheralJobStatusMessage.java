/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents;

import java.time.Instant;
import java.util.List;
import java.util.stream.Collectors;
import org.opentcs.data.peripherals.PeripheralJob;
import org.opentcs.data.peripherals.PeripheralJob.State;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.PeripheralOperationDescription;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.Property;

/**
 * A status message containing information about a peripheral job.
 */
public class PeripheralJobStatusMessage
    extends StatusMessage {

  private String name;

  private String reservationToken;

  private String relatedVehicle;

  private String relatedTransportOrder;

  private PeripheralOperationDescription peripheralOperation;

  private State state;

  private Instant creationTime;

  private Instant finishedTime;

  private List<Property> properties;

  /**
   * Creates a new instance.
   */
  public PeripheralJobStatusMessage() {
  }

  @Override
  public PeripheralJobStatusMessage setSequenceNumber(long sequenceNumber) {
    return (PeripheralJobStatusMessage) super.setSequenceNumber(sequenceNumber);
  }

  @Override
  public PeripheralJobStatusMessage setCreationTimeStamp(Instant creationTimeStamp) {
    return (PeripheralJobStatusMessage) super.setCreationTimeStamp(creationTimeStamp);
  }

  public String getName() {
    return name;
  }

  public PeripheralJobStatusMessage setName(String name) {
    this.name = name;
    return this;
  }

  public String getReservationToken() {
    return reservationToken;
  }

  public PeripheralJobStatusMessage setReservationToken(String reservationToken) {
    this.reservationToken = reservationToken;
    return this;
  }

  public String getRelatedVehicle() {
    return relatedVehicle;
  }

  public PeripheralJobStatusMessage setRelatedVehicle(String relatedVehicle) {
    this.relatedVehicle = relatedVehicle;
    return this;
  }

  public String getRelatedTransportOrder() {
    return relatedTransportOrder;
  }

  public PeripheralJobStatusMessage setRelatedTransportOrder(String relatedTransportOrder) {
    this.relatedTransportOrder = relatedTransportOrder;
    return this;
  }

  public PeripheralOperationDescription getPeripheralOperation() {
    return peripheralOperation;
  }

  public PeripheralJobStatusMessage setPeripheralOperation(
      PeripheralOperationDescription peripheralOperation) {
    this.peripheralOperation = peripheralOperation;
    return this;
  }

  public State getState() {
    return state;
  }

  public PeripheralJobStatusMessage setState(State state) {
    this.state = state;
    return this;
  }

  public Instant getCreationTime() {
    return creationTime;
  }

  public PeripheralJobStatusMessage setCreationTime(Instant creationTime) {
    this.creationTime = creationTime;
    return this;
  }

  public Instant getFinishedTime() {
    return finishedTime;
  }

  public PeripheralJobStatusMessage setFinishedTime(Instant finishedTime) {
    this.finishedTime = finishedTime;
    return this;
  }

  public List<Property> getProperties() {
    return properties;
  }

  public PeripheralJobStatusMessage setProperties(List<Property> properties) {
    this.properties = properties;
    return this;
  }

  public static PeripheralJobStatusMessage fromPeripheralJob(PeripheralJob job,
                                                             long sequenceNumber) {
    return fromPeripheralJob(job, sequenceNumber, Instant.now());
  }

  public static PeripheralJobStatusMessage fromPeripheralJob(PeripheralJob job,
                                                             long sequenceNumber,
                                                             Instant creationTimestamp) {
    PeripheralJobStatusMessage message = new PeripheralJobStatusMessage();
    message.setSequenceNumber(sequenceNumber);
    message.setCreationTimeStamp(creationTimestamp);

    message.setName(job.getName());
    message.setReservationToken(job.getReservationToken());
    if (job.getRelatedVehicle() != null) {
      message.setRelatedVehicle(job.getRelatedVehicle().getName());
    }
    if (job.getRelatedTransportOrder() != null) {
      message.setRelatedTransportOrder(job.getRelatedTransportOrder().getName());
    }
    message.setPeripheralOperation(
        PeripheralOperationDescription.fromPeripheralOperation(job.getPeripheralOperation())
    );
    message.setState(job.getState());
    message.setCreationTime(job.getCreationTime());
    message.setFinishedTime(job.getFinishedTime());
    message.setProperties(job.getProperties().entrySet().stream()
        .map(entry -> new Property(entry.getKey(), entry.getValue()))
        .collect(Collectors.toList()));

    return message;
  }
}
