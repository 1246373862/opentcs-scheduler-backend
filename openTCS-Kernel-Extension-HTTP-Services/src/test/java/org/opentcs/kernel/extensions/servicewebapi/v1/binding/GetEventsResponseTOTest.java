/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.kernel.extensions.servicewebapi.v1.binding;

import java.time.Instant;
import java.util.List;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.data.model.Vehicle;
import org.opentcs.data.peripherals.PeripheralJob;
import org.opentcs.data.peripherals.PeripheralOperation;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents.OrderStatusMessage;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents.PeripheralJobStatusMessage;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents.VehicleStatusMessage;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.DestinationState;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.PeripheralOperationDescription;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.Property;

/**
 * Unit tests for {@link GetEventsResponseTO}.
 */
public class GetEventsResponseTOTest {

  private JsonBinder jsonBinder;

  @BeforeEach
  public void setUp() {
    jsonBinder = new JsonBinder();
  }

  @Test
  public void jsonSample() {
    GetEventsResponseTO to
        = new GetEventsResponseTO()
            .setTimeStamp(Instant.EPOCH)
            .setStatusMessages(
                List.of(
                    createVehicleStatusMessage(0),
                    createOrderStatusMessage(1),
                    createPeripheralJobStatusMessage(2)
                )
            );

    Approvals.verify(jsonBinder.toJson(to));
  }

  private VehicleStatusMessage createVehicleStatusMessage(long sequenceNo) {
    return new VehicleStatusMessage()
        .setSequenceNumber(sequenceNo)
        .setCreationTimeStamp(Instant.EPOCH)
        .setVehicleName("some-vehicle")
        .setTransportOrderName("some-transport-order")
        .setPosition("some-point")
        .setPrecisePosition(new VehicleStatusMessage.PrecisePosition(1, 2, 3))
        .setPaused(false)
        .setState(Vehicle.State.IDLE)
        .setProcState(Vehicle.ProcState.IDLE)
        .setAllocatedResources(
            List.of(
                List.of("some-path", "some-point"),
                List.of("some-other-path", "some-other-point")
            )
        )
        .setClaimedResources(
            List.of(
                List.of("some-path", "some-point"),
                List.of("some-other-path", "some-other-point")
            )
        );
  }

  private OrderStatusMessage createOrderStatusMessage(long sequenceNo) {
    return new OrderStatusMessage()
        .setSequenceNumber(sequenceNo)
        .setCreationTimeStamp(Instant.EPOCH)
        .setOrderName("some-order")
        .setProcessingVehicleName("some-vehicle")
        .setOrderState(OrderStatusMessage.OrderState.BEING_PROCESSED)
        .setDestinations(
            List.of(
                new DestinationState()
                    .setLocationName("some-location")
                    .setOperation("some-operation")
                    .setState(DestinationState.State.TRAVELLING)
                    .setProperties(
                        List.of(
                            new Property("some-key", "some-value"),
                            new Property("some-other-key", "some-other-value")
                        )
                    )
            )
        )
        .setProperties(
            List.of(
                new Property("some-key", "some-value"),
                new Property("some-other-key", "some-other-value")
            )
        );
  }

  private PeripheralJobStatusMessage createPeripheralJobStatusMessage(long sequenceNo) {
    return new PeripheralJobStatusMessage()
        .setSequenceNumber(sequenceNo)
        .setCreationTimeStamp(Instant.EPOCH)
        .setName("some-peripheral-job")
        .setReservationToken("some-token")
        .setRelatedVehicle("some-vehicle")
        .setRelatedTransportOrder("some-order")
        .setPeripheralOperation(
            new PeripheralOperationDescription()
                .setOperation("some-operation")
                .setLocationName("some-location")
                .setExecutionTrigger(PeripheralOperation.ExecutionTrigger.AFTER_ALLOCATION)
                .setCompletionRequired(true)
        )
        .setState(PeripheralJob.State.BEING_PROCESSED)
        .setCreationTime(Instant.EPOCH)
        .setFinishedTime(Instant.MAX)
        .setProperties(
            List.of(
                new Property("some-key", "some-value"),
                new Property("some-other-key", "some-other-value")
            )
        );
  }
}
