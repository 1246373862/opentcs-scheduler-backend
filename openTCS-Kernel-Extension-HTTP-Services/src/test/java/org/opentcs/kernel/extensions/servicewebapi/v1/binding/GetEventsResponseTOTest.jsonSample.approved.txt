{
  "timeStamp" : "1970-01-01T00:00:00Z",
  "statusMessages" : [ {
    "type" : "Vehicle",
    "sequenceNumber" : 0,
    "creationTimeStamp" : "1970-01-01T00:00:00Z",
    "vehicleName" : "some-vehicle",
    "transportOrderName" : "some-transport-order",
    "position" : "some-point",
    "precisePosition" : {
      "x" : 1,
      "y" : 2,
      "z" : 3
    },
    "paused" : false,
    "state" : "IDLE",
    "procState" : "IDLE",
    "allocatedResources" : [ [ "some-path", "some-point" ], [ "some-other-path", "some-other-point" ] ],
    "claimedResources" : [ [ "some-path", "some-point" ], [ "some-other-path", "some-other-point" ] ]
  }, {
    "type" : "TransportOrder",
    "sequenceNumber" : 1,
    "creationTimeStamp" : "1970-01-01T00:00:00Z",
    "orderName" : "some-order",
    "processingVehicleName" : "some-vehicle",
    "orderState" : "BEING_PROCESSED",
    "destinations" : [ {
      "locationName" : "some-location",
      "operation" : "some-operation",
      "state" : "TRAVELLING",
      "properties" : [ {
        "key" : "some-key",
        "value" : "some-value"
      }, {
        "key" : "some-other-key",
        "value" : "some-other-value"
      } ]
    } ],
    "properties" : [ {
      "key" : "some-key",
      "value" : "some-value"
    }, {
      "key" : "some-other-key",
      "value" : "some-other-value"
    } ]
  }, {
    "type" : "PeripheralJob",
    "sequenceNumber" : 2,
    "creationTimeStamp" : "1970-01-01T00:00:00Z",
    "name" : "some-peripheral-job",
    "reservationToken" : "some-token",
    "relatedVehicle" : "some-vehicle",
    "relatedTransportOrder" : "some-order",
    "peripheralOperation" : {
      "operation" : "some-operation",
      "locationName" : "some-location",
      "executionTrigger" : "AFTER_ALLOCATION",
      "completionRequired" : true
    },
    "state" : "BEING_PROCESSED",
    "creationTime" : "1970-01-01T00:00:00Z",
    "finishedTime" : "+1000000000-12-31T23:59:59.999999999Z",
    "properties" : [ {
      "key" : "some-key",
      "value" : "some-value"
    }, {
      "key" : "some-other-key",
      "value" : "some-other-value"
    } ]
  } ]
}